File: api-server/src/models/personal.rs
use chrono::NaiveDate;
use serde::Deserialize;
use serde::Serialize;
use sqlx::FromRow;
use validator::Validate;
use validator::ValidationError;

/// Valida que el DNI tenga exactamente 8 dígitos numéricos.
pub fn es_dni_valido(dni: &str) -> Result<(), ValidationError> {
    if dni.len() == 8 && dni.chars().all(|c| c.is_ascii_digit()) {
        Ok(())
    } else {
        let mut err = ValidationError::new("dni_invalido");
        err.message = Some("Debe tener exactamente 8 dígitos numéricos".into());
        Err(err)
    }
}

#[derive(Serialize)]
pub struct Persona {
    pub nombre: Option<String>,
    pub dni: String,
    pub estado: Option<String>,
    pub sexo: Option<String>,
}

#[derive(Serialize, Deserialize, Validate)]
pub struct Perfil {
    #[validate(custom(function = "es_dni_valido"))]
    pub dni: String,
    pub nombre: Option<String>,
    pub telf: Option<String>,
    pub direccion: Option<String>,
    pub email: Option<String>,
    pub ruc: Option<String>,
    pub nacimiento: NaiveDate,
    pub sexo: Option<String>,
    pub region: Option<String>,
    pub distrito: Option<String>,
}

#[derive(Serialize, Deserialize)]
pub struct Vinculos {
    pub id: i32,
    pub dni: String,
    pub doc_ingreso: Option<String>,
    pub numero_doc_ingreso: Option<String>,
    pub descrip_ingreso: Option<String>,
    pub fecha_ingreso: Option<NaiveDate>,
    pub area: Option<String>,
    pub cargo: String,
    pub regimen: Option<String>,
    pub sueldo: Option<f64>,
    pub codigo: Option<String>,
    pub cargo_estructural: Option<String>,
    pub grupo_ocupacional: Option<String>,
    pub estado: String,
    pub doc_salida: Option<String>,
    pub descrip_salida: Option<String>,
    pub fecha_salida: Option<NaiveDate>,
    pub numero_doc_salida: Option<String>,
    pub sindicato: Option<String>,
    pub tipo_evento: Option<String>,
    pub estado_evento: Option<String>,
    pub doc_evento_tipo: Option<String>,
    pub numero_doc_evento: Option<i32>,
    pub fecha_evento: Option<NaiveDate>,
    pub id_evento: Option<i32>,
}

#[derive(Debug, Serialize, Deserialize, FromRow, Validate)]
pub struct Documento {
    pub id: Option<i32>,
    #[serde(rename = "tipoDocumento")]
    pub tipo: Option<String>,
    #[serde(rename = "numeroDocumento")]
    pub numero: Option<i32>,
    #[serde(rename = "añoDocumento")]
    pub año: Option<i32>,
    #[validate(length(min = 1, message = "La fecha es requerida"))]
    pub fecha: String,
    #[serde(rename = "fechaValida")]
    pub fecha_valida: Option<String>,
    pub conv: Option<i64>,
    #[validate(length(min = 1, message = "La descripción es requerida"))]
    pub descripcion: String,
    pub funcion: Option<i64>,
}

#[derive(Debug, Serialize, Deserialize, FromRow, Validate)]
pub struct DatosBancarios {
    pub id: i32,
    #[validate(length(
        min = 6,
        max = 30,
        message = "Número de cuenta inválido (6-30 caracteres)"
    ))]
    pub numero_cuenta: String,
    pub tipo_cuenta: Option<String>,
    pub cci: Option<String>,
    pub banco: String,
    pub dni: String,
    pub estado: i8,
}

#[derive(Debug, Serialize, Deserialize, FromRow, Validate)]
pub struct DatosBancariosResponse {
    #[validate(length(
        min = 6,
        max = 30,
        message = "Número de cuenta inválido (6-30 caracteres)"
    ))]
    pub numero_cuenta: String,
    #[validate(length(min = 1, message = "El tipo de cuenta es requerido"))]
    pub tipo_cuenta: String,
    pub cci: Option<String>,
    pub banco: i32,
    pub estado: i8,
    #[validate(custom(function = "es_dni_valido"))]
    pub dni: String,
}

#[derive(Debug, Serialize, Deserialize, FromRow, Validate)]
pub struct GradoAcademico {
    pub id: i32,
    #[validate(length(min = 2, message = "Mínimo 2 caracteres"))]
    pub profesion: String,
    #[validate(length(min = 2, message = "Mínimo 2 caracteres"))]
    pub universidad: String,
    pub colegiatura: Option<String>,
    #[validate(length(min = 2, message = "Mínimo 2 caracteres"))]
    pub nivel_academico: String,
    #[validate(length(min = 1, message = "La abreviatura es requerida"))]
    pub abrv: String,
    #[validate(custom(function = "es_dni_valido"))]
    pub dni: String,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct VinculosSindicato {
    pub id_vinculo: i32,
    pub dni: String,
}

#[derive(Debug, Serialize, Deserialize, FromRow, Validate)]
pub struct DocumentoSindicato {
    pub id: Option<i32>,
    #[serde(rename = "tipoDocumento")]
    pub tipo: Option<String>,
    #[serde(rename = "numeroDocumento")]
    pub numero: Option<i32>,
    #[serde(rename = "añoDocumento")]
    pub año: Option<i32>,
    #[validate(length(min = 1, message = "La fecha es requerida"))]
    pub fecha: String,
    #[serde(rename = "fechaValida")]
    pub fecha_valida: Option<String>,
    #[validate(length(min = 1, message = "La descripción es requerida"))]
    pub descripcion: String,
    pub sindicato: i32,
    pub vinculos: Vec<VinculosSindicato>,
}

#[derive(Debug, Serialize, Deserialize, FromRow, Validate)]
pub struct LegajoPersonal {
    pub id: i32,
    #[validate(length(min = 2, message = "Mínimo 2 caracteres"))]
    pub persona: String,
    pub dni: Option<String>,
    pub fecha: Option<String>,
    pub estado: Option<String>,
    pub descrip: Option<String>,
    pub nuevo: i32,
    pub user: Option<i32>,
}

#[derive(Debug, Serialize, Deserialize, FromRow)]
pub struct AsistenciaVw {
    pub dni: String,
    pub hora: Option<String>,
    pub fecha: Option<NaiveDate>,
}

#[derive(Debug, Serialize, Deserialize, FromRow, Validate)]
pub struct ContactoEmergencia {
    #[validate(custom(function = "es_dni_valido"))]
    pub persona_dni: String,
    #[validate(length(min = 2, message = "Mínimo 2 caracteres"))]
    pub nombre: String,
    #[validate(length(min = 2, message = "Mínimo 2 caracteres"))]
    pub relacion: String,
    pub telefono: Option<String>,
}

#[derive(Serialize, Deserialize, Validate)]
pub struct PerfilInput {
    #[validate(custom(function = "es_dni_valido"))]
    pub dni: String,
    #[validate(length(min = 1, message = "El apellido materno es requerido"))]
    pub amaterno: String,
    #[validate(length(min = 1, message = "El apellido paterno es requerido"))]
    pub apaterno: String,
    #[validate(length(min = 1, message = "El nombre es requerido"))]
    pub nombre: String,
    pub telf: Option<String>,
    pub direccion: Option<String>,
    pub email: Option<String>,
    pub ruc: Option<String>,
    pub nacimiento: NaiveDate,
    pub sexo: Option<String>,
    pub region: Option<String>,
    pub distrito: Option<String>,
}

#[derive(Serialize, Deserialize, Validate)]
pub struct NuevoVinculo {
    #[validate(nested)]
    pub personal: PerfilInput,
    #[validate(length(min = 1, message = "El código de plaza es requerido"))]
    pub airshp: String,
    #[validate(nested)]
    pub documento: Documento,
    #[validate(range(min = 1, message = "ID de régimen inválido"))]
    pub regimen: i32,
    #[validate(range(min = 1, message = "ID de cargo inválido"))]
    pub cargo: i32,
    #[validate(range(min = 1, message = "ID de área inválido"))]
    pub area: i32,
    #[validate(range(min = 0.0, message = "El sueldo debe ser mayor o igual a 0"))]
    pub sueldo: f64,
}

#[derive(Debug, Serialize, Deserialize, Validate)]
pub struct EventoVinculoPayload {
    pub id: Option<i32>,
    #[validate(range(min = 1, message = "ID de vínculo inválido"))]
    pub vinculo_id: i32,
    #[validate(length(min = 1, message = "El tipo de evento es requerido"))]
    pub tipo_evento: String,
    pub nueva_area_id: Option<i32>,
    pub documento_inicio: Option<Documento>,
    pub documento_salida: Option<Documento>,
    pub estado: Option<String>,
}

#[derive(Deserialize, Validate)]
pub struct DeleteEventoVinculoPayload {
    #[validate(range(min = 1, message = "ID de evento inválido"))]
    pub id: i32,
}

File: crm/src/services/personal.ts
import { api, getApiBaseUrl } from './api'

export interface PersonalPerfil {
  dni: string
  nombre: string
  telf: string | null
  direccion: string | null
  email: string | null
  ruc: string | null
  nacimiento: string | null
  sexo: string | null
  region: string | null
  distrito: string | null
}

export interface PersonalBanco {
  id: number
  numero_cuenta: string
  tipo_cuenta: string
  cci: string
  banco: string
  estado: number
  dni?: string
}

export interface PersonalGrado {
  id: number
  profesion: string
  universidad: string
  nivel_academico: string
  abrv: string
  dni: string
  fecha: string
}

export interface PersonalContacto {
  persona_dni: string
  nombre: string
  relacion: string
  telefono: string
}

export interface PersonalVinculo {
  id: number
  dni: string
  area: string
  cargo: string
  regimen: string
  sueldo: number
  codigo: string
  estado: string
  fecha_ingreso: string
  fecha_salida: string | null
  sindicato: string | null
  tipo_evento: string | null
  estado_evento: string | null
  id_evento: number | null
  doc_ingreso_id?: number | null
  doc_ingreso?: string | null
  numero_doc_ingreso?: string | null
  descrip_ingreso?: string | null
  cargo_estructural?: string | null
  grupo_ocupacional?: string | null
  doc_salida_id?: number | null
  doc_salida?: string | null
  descrip_salida?: string | null
  numero_doc_salida?: string | null
  doc_evento_id?: number | null
  doc_evento_tipo?: string | null
  numero_doc_evento?: string | null
  fecha_evento?: string | null
}

export interface PersonalArchivo {
  id: number
  documento_id: number | null
  dni_asociado: string
  original_name: string
  file_hash: string
  extension: string
  external_url: string | null
  usuario_subida: string
  fecha_subida: string
  size?: string
  access?: string
  access_members?: number
}

export interface PersonalDocumento {
  id: number
  sigla: string
  fecha: string
  descripcion: string
}

export interface PersonalSearchResult {
  nombre: string
  dni: string
  estado: string
  sexo: string
}

export function getFileDownloadUrl(hash: string): string {
  const baseUrl = getApiBaseUrl()
  return `${baseUrl}/fileserver/${hash}`
}

export async function fetchPersonalPerfil(dni: string): Promise<PersonalPerfil> {
  return await api<PersonalPerfil>(`/personal/perfil/${dni}`)
}

export async function fetchPersonalBanco(dni: string): Promise<PersonalBanco | null> {
  try {
    return await api<PersonalBanco | null>(`/personal/banco/${dni}`)
  } catch {
    return null
  }
}

export async function fetchPersonalGrados(dni: string): Promise<PersonalGrado[]> {
  try {
    return await api<PersonalGrado[]>(`/personal/grado/${dni}`)
  } catch {
    return []
  }
}

export async function fetchPersonalContacto(dni: string): Promise<PersonalContacto | null> {
  try {
    return await api<PersonalContacto | null>(`/personal/contacto/${dni}`)
  } catch {
    return null
  }
}

export async function fetchPersonalVinculos(dni: string): Promise<PersonalVinculo[]> {
  try {
    return await api<PersonalVinculo[]>(`/personal/vinculos/${dni}`)
  } catch {
    return []
  }
}

export async function fetchPersonalArchivos(dni: string): Promise<PersonalArchivo[]> {
  try {
    return await api<PersonalArchivo[]>(`/fileserver/archivos_por_dni/${dni}`)
  } catch {
    return []
  }
}

export async function fetchPersonalDocumentos(dni: string): Promise<PersonalDocumento[]> {
  try {
    return await api<PersonalDocumento[]>(`/fileserver/documentos/${dni}`)
  } catch {
    return []
  }
}

export async function buscarTrabajadores(nombre: string): Promise<PersonalSearchResult[]> {
  try {
    return await api<PersonalSearchResult[]>('/personal/buscar', {
      query: { nombre },
    })
  } catch {
    return []
  }
}

export async function updatePersonalPerfil(perfil: PersonalPerfil): Promise<boolean> {
  await api('/personal/editar_por_dni', {
    method: 'PUT',
    body: perfil,
  })
  return true
}

export function getPersonalAvatarUrl(dni: string): string {
  const baseUrl = getApiBaseUrl()
  return `${baseUrl}/personal/avatar/${dni}`
}

export function formatMoneda(val: number | null | undefined): string {
  if (val === null || val === undefined) return 'S/ 0.00'
  return new Intl.NumberFormat('es-PE', { style: 'currency', currency: 'PEN' }).format(val)
}

export type VinculoStatusType = 'success' | 'warning' | 'normal'

export function getVinculoStatusType(v: PersonalVinculo | null | undefined): VinculoStatusType {
  if (!v) return 'normal'
  if (v.estado?.toLowerCase() === 'activo') return 'success'
  const hasDocSalida = Boolean(
    (v.doc_salida && v.doc_salida.trim()) ||
    (v.numero_doc_salida && v.numero_doc_salida.trim())
  )
  return hasDocSalida ? 'normal' : 'warning'
}
